File: Array/fanbei/0702/c/solution07_02.c
#include <stdlib.h>
#include <stdio.h>

/**
 * 有序链表转换二叉搜索树给定一个单链表，其中的元素按升序排序，将其转换为高度平衡的二叉搜索树,
 * 本题中，一个高度平衡二叉树是指一个二叉树每个节点的左右两个子树的高度差的绝对值不超过1.
 * 这个题，还是有点不理解，后面再屡屡
 */

struct ListNode{
    int val;
    struct ListNode* next;
};

struct TreeNode{
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
};

int sizeofListNode(struct ListNode *head) {
    struct ListNode *tmp = head;
    int size = 0;
    while (tmp != NULL) {
        size += 1;
        tmp = tmp->next;
    }
    return size;
}

struct ListNode *tmpHead;

struct TreeNode *convertListToBST(int left, int right) {
    if (left > right) {
        return NULL;
    }

    int mid = (left + right) / 2;
    struct TreeNode *leftTreeNode = convertListToBST(left, mid - 1);
    struct TreeNode *root = malloc(sizeof(struct TreeNode));
    root->val = tmpHead->val;
    root->left = leftTreeNode;
    tmpHead = tmpHead->next;
    root->right = convertListToBST(mid + 1, right);
    return root;
}

struct TreeNode *sortedListToBST(struct ListNode *head) {
    tmpHead = head;
    int size = sizeofListNode(head);
    return convertListToBST(0, size - 1);
}

File: Array/fanbei/0702/java/Solution07_02.java
package com.example.lsn37_dagger2_subcomponent;

/**
 * 109. 有序链表转换二叉搜索树
 * 别人的实例，还得再理解一下。
 */
public class Solution07_02 {

    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
            left = null;
            right = null;
        }
    }

    public TreeNode sortedListToBST(ListNode head) {

        if (head == null) {
            return null;
        }
        ListNode mid = findMid(head);
        TreeNode root = new TreeNode(mid.val);
        if (head == mid) {
            return root;
        }
        root.left = sortedListToBST(head);
        root.right = sortedListToBST(mid.next);
        return root;
    }

    private ListNode findMid(ListNode head) {
        if (head == null) {
            return null;
        }
        ListNode fast = head;
        ListNode low = head;
        ListNode pre = null;
        while (fast != null && fast.next != null) {
            pre = low;
            fast = fast.next.next;
            low = low.next;
        }

        if (pre != null) {
            pre.next = null;
        }
        return low;
    }

}
